h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java

```diff
@@ -44,90 +45,95 @@ public ValFrame exec(Val[] args) {
     // This is the sort then MRTask method.
     // Create the target Frame
     // Now sort on the index key, result is that unique keys will be localized
-    Frame fr2 = fr.sort(new int[]{0});
+    Frame fr2 = fr.sort(new int[]{indexIdx});
     final long[] classes = new VecUtils.CollectDomain().doAll(fr.vec(colIdx)).domain();
-    final int nClass = fr.vec(colIdx).isNumeric() ? classes.length : fr.vec(colIdx).domain().length;
-    String[] header = (String[]) ArrayUtils.addAll(new String[]{index}, fr.vec(colIdx).domain());
-    Frame initialPassPreSort = new pivotTask(fr2,index,column,value)
+    final int nClass = (fr.vec(colIdx).isNumeric() || fr.vec(colIdx).isTime()) ? classes.length : fr.vec(colIdx).domain().length;
+    String[] header = null;
+    if (fr.vec(colIdx).isNumeric()) {
+      header = (String[]) ArrayUtils.addAll(new String[]{index}, Arrays.toString(classes).split("[\\[\\]]")[1].split(", "));
+    } else if (fr.vec(colIdx).isTime()) {
+      header = new String[nClass];
+      for (int i=0;i<nClass;i++) header[i] = (new DateTime(classes[i], DateTimeZone.UTC)).toString();
+    } else {
+      header = (String[]) ArrayUtils.addAll(new String[]{index}, fr.vec(colIdx).domain());
+    }
+
+    Frame initialPass = new pivotTask(fr2.find(index),fr2.find(column),fr2.find(value),classes)
       .doAll(nClass+1, Vec.T_NUM, fr2)
       .outputFrame(null, header, null);
     fr2.delete();
-    Frame initialPass = initialPassPreSort.sort(new int[]{0});
-    initialPassPreSort.delete();
-    // Collapse identical index rows even when index crosses over chunk boundaries
-    Frame secondPass = new pivotCleanup()
-      .doAll(nClass+1,Vec.T_NUM,initialPass)
-      .outputFrame(null,header,null);
-
-    initialPass.delete();
-    Frame result = new Frame(secondPass.vec(0).makeCopy(null,fr.vec(indexIdx).get_type()));
+    Frame result = new Frame(initialPass.vec(0).makeCopy(fr.vec(indexIdx).domain(),fr.vec(indexIdx).get_type()));
```

Comment:
Im passing the domain now so that if the index is categorical, it maintains the labels